### PR TITLE
Fix the inconsistency of loss calculation between PT/TF XLNetLMHeadModel

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

The loss calculation in `XLNetLMHeadModel` doesn't cut the logits/labels (unlike other models say `BertLMHeadModel`).
However, `TFXLNetLMHeadModel` works as other TF Causal LM models, and cut the logits.
This causes the loss difference higher than `4e-2` sometimes.

I believe `XLNet` works somehow differently from the usual causal LM models, and the provided labels and computed logits shouldn't be cut, as  `XLNetLMHeadModel` does.

This PR fixes this inconsistency.

### CODE REVIEW

**Comment by LysandreJik** [comment]
Could @Rocketknight1 or @gante give this a quick look?

**Comment by Rocketknight1** [comment]
This looks good! Should it be folded into PR #15256 though, or do you think it's better separate?

**Comment by ydshieh** [comment]
> This looks good! Should it be folded into PR #15256 though, or do you think it's better separate?

Hi, @Rocketknight1 

 #15256  we still cut the logits to compute the loss, but we will return the complete logits
But for this PR, we want to **use the complete logits to compute** the loss (and then return the complete one).
(that's my understanding)

They are different, and would be better to be separate PRs. There will be a conflict once one version is merged, but I can take care of resolving it.

**Comment by gante** [comment]
Nice finding @ydshieh!

I agree with @ydshieh, these are two separate issues -- one is general to causal models, this one pertains to the TF/PT mismatch in XLNet.

File: tests/test_modeling_tf_xlnet.py
Comment (gante) [inline tests/test_modeling_tf_xlnet.py:412]: If I understand correctly, this is the difference between the original test and the overwritten one we have here. We can also rewrite this part in the original test into something like the snippet below, removing the need to overwrite the test (which might prevent issues down the road, if we need to touch this test again).

```python
if model.__class__ in get_values(TF_MODEL_FOR_CAUSAL_LM_MAPPING):
    # `TFXLNetLMHeadModel` doesn't cut logits/labels
    if not 'TFXLNetLMHeadModel' in str(model.__class__):
        loss_size = loss_size - self.model_tester.batch_size
```

@Rocketknight1 WDYT?

**Review by gante — COMMENTED** [review]
LGTM, let's discuss the comment I added before merging :)

File: tests/test_modeling_tf_xlnet.py
Comment (ydshieh) [reply tests/test_modeling_tf_xlnet.py:412]: I did it this way simply because from a previous experience (another PR), @sgugger suggested me to move this kind of model-specific logic to its own test. I think both approaches have their own pros/cons, but I don't have strong opinion on it for this case.

cc @sgugger to see if he has opinion on it.

Comment (sgugger) [reply tests/test_modeling_tf_xlnet.py:412]: Yes, we try to avoid too much custom logic in the common tests, otherwise they become too quickly unreadable with 100 model architectures, each with its own need.

Comment (gante) [reply tests/test_modeling_tf_xlnet.py:412]: Cool, makes sense 👍 @sgugger Do we have some mechanism to ensure we don't forget to propagate changes to custom tests?

Comment (sgugger) [reply tests/test_modeling_tf_xlnet.py:412]: Sadly not.

**Review by gante — APPROVED** [review]
🚀 
(please wait for @Rocketknight1's approval before merging)

**Review by Rocketknight1 — APPROVED** [review]

**Comment by Rocketknight1** [comment]
Sorry for the delay! I looked at this earlier in the week and thought it looked good, and didn't realize you were waiting on me now. I'm happy to merge at this point - @ydshieh is there anything else you want to add, or should I go ahead?

**Comment by ydshieh** [comment]
> Sorry for the delay! I looked at this earlier in the week and thought it looked good, and didn't realize you were waiting on me now. I'm happy to merge at this point - @ydshieh is there anything else you want to add, or should I go ahead?

You can go ahead :-) Thank you!

**Comment by gante** [comment]
Merging then 👍 
Thank you for the PR, @ydshieh!

**Pull request merged**
